plugins/autolink/plugin.js

```diff
@@ -6,45 +6,76 @@
 ( function() {
 	'use strict';
 
-	// Regex by Imme Emosol.
-	var validUrlRegex = /^(https?|ftp):\/\/(-\.)?([^\s\/?\.#]+\.?)+(\/[^\s]*)?[^\s\.,]$/ig,
-		// Regex by (https://www.w3.org/TR/html5/forms.html#valid-e-mail-address).
-		validEmailRegex = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/g,
-		doubleQuoteRegex = /"/g;
+	var doubleQuoteRegex = /"/g;
 
 	CKEDITOR.plugins.add( 'autolink', {
-		requires: 'clipboard',
+		requires: 'clipboard,textmatch',
 
 		init: function( editor ) {
+			var urlTemplate = new CKEDITOR.template( '<a href="{link}">{text}</a>' ),
+				emailTemplate = new CKEDITOR.template( '<a href="mailto:{link}">{text}</a>' );
+
 			editor.on( 'paste', function( evt ) {
-				var data = evt.data.dataValue;
+				// Pasted text may be encoded. Decode it so we will be able to compare its length with match.
+				var text = CKEDITOR.tools.htmlDecode( evt.data.dataValue );
 
-				if ( evt.data.dataTransfer.getTransferType( editor ) == CKEDITOR.DATA_TRANSFER_INTERNAL ) {
-					return;
+				// Attach afterPaste event here to avoid race condition between events.
+				editor.once( 'afterPaste', function() {
+					autolink( function( matched ) {
+						return matched && matched.text.length == text.length;
+					} );
+				} );
+			} );
+
+			// IE has its own link completion and we don't want to interfere with it.
+			if ( CKEDITOR.env.ie && !CKEDITOR.env.edge ) {
+				return;
+			}
+
+			editor.on( 'contentDom', function() {
+				var commitKeystrokes = editor.config.autolink_commitKeystrokes || CKEDITOR.config.autolink_commitKeystrokes;
+				editor.editable().on( 'keydown', function( evt ) {
```

Comment:
We have dedicated event for key handling in editor, [`key`](https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#event-key).